Common/Data/LeanDataWriter.cs

```diff
@@ -172,328 +213,99 @@ public void DownloadAndSave(IBrokerage brokerage, List<Symbol> symbols, DateTime
-
-                    if (zip.Count > 0)
-                    {
-                        zip.Save();
+                        var time = DateTime.ParseExact(line.AsSpan(0, DateFormat.TwelveCharacter.Length), DateFormat.TwelveCharacter, CultureInfo.InvariantCulture);
+                        rows[time] = line;
                     }
                 }
 
-                date = date.AddDays(1);
+                return true;
             }
         }
 
         /// <summary>
-        /// Write out the data in LEAN format (minute, second or tick resolutions)
+        /// Write this file to disk with the given data.
         /// </summary>
-        /// <param name="source">IEnumerable source of the data: sorted from oldest to newest.</param>
-        /// <remarks>This function overwrites existing data files</remarks>
-        private void WriteMinuteOrSecondOrTick(IEnumerable<BaseData> source)
+        /// <param name="filePath">The full path to the new file</param>
+        /// <param name="data">The data to write as a list of dates and strings</param>
+        /// <param name="date">The date the data represents</param>
+        /// <remarks>The reason we have the data as IEnumerable(DateTime, string) is to support
+        /// a generic write that works for all resolutions. In order to merge in hour/daily case I need the
+        /// date of the data to correctly merge the two. In order to support writing ticks I need to allow
+        /// two data points to have the same time. Thus I cannot use a single list of just strings nor
+        /// a sorted dictionary of DateTimes and strings. Only upon reaching the merge case will I sort and
+        /// merge, or just extract string data from the tuple.</remarks>
+        private void WriteFile(string filePath, IEnumerable<(DateTime, string)> data, DateTime date)
         {
-            var lines = new List<string>();
-            var lastTime = new DateTime();
-
-            // Loop through all the data and write to file as we go
-            foreach (var data in source)
+            // Generate this csv entry name
+            var entryName = LeanData.GenerateZipEntryName(_symbol, date, _resolution, _tickType);
+            
+            // Check disk once for this file ahead of time, reuse where possible
+            var fileExists = File.Exists(filePath);
+
+            // Handle merging of files
+            // Only merge on files with hour/daily resolution, that exist, and can be loaded
+            string finalData;
+            if (_resolution >= Resolution.Hour && fileExists && TryLoadFile(filePath, entryName, out var rows))
             {
-                // Ensure the data is sorted
-                if (data.Time < lastTime) throw new Exception("The data must be pre-sorted from oldest to newest");
+                // Load this data into a sorted dictionary so we may commence the merge
+                var sortedNewData = data.ToImmutableSortedDictionary(x => x.Item1, x => x.Item2);
```

Comment:
Not sure `ToImmutableSortedDictionary` is required, can directly iterate over `data` bellow?

Reply:
Oh nice! Yes that would work, I'm glad too because I really didn't like how it was.